00-config-refactor.md

```diff
@@ -0,0 +1,102 @@
+# Proposal: Config Refactor
+
+## Background
+
+For historical reasons, our SDK has the following problems:
+
+- SDK's config path is hardcoded to `~/.config/qingstor.yaml`, if not exist, SDK will create automatically.
+- There is not way for user to change SDK's config path during runtime. This enforces all tools to implement their own
+  SDK loading logic.
+- SDK's assume all user have access key pairs which forbids anonymous API call.
+- SDK use virtual host style to check bucket but use path style for API call.
+- SDK can't control IPv4/IPv6 logic
+
+## Proposal
+
+So I propose to do following config refactors:
+
+### Environment First Config
+
+Instead of hardcoded `~/.config/qingstor.yaml`, we will read config path from system environment
+variable `QINGSTOR_CONFIG_PATH`.
+
+If `QINGSTOR_CONFIG_PATH` is not exists, we will fallback to `~/.config/qingstor.yaml`.
+
+And all SDKs should allow reading config from envs:
+
+```shell
+export QINGSTOR_ACCESS_KEY_ID=example_access_key_id
+export QINGSTOR_SECRET_ACCESS_KEY=example_secret_access_key
+export QINGSTOR_XYZ=example_xyz
+```
+
+For every config, we should check via following order:
+
+- The specific env: `QINGSTOR_ACCESS_KEY_ID`
+- Config loading from `QINGSTOR_CONFIG_PATH`
+- If no default value, keep empty, or use the default value.
+
+All qingstor native tools could use the same logic here.
```

Comment:
So all the tools (like `qsctl` and `qscamel`) do not need to handle config as an independent flag? e.g. `-c` or `--config`

Reply:
They don't need to care about sdk config any more. 

But some tools may need `-c` or `--config` to handle they own config.